### PR TITLE
polkit-agent: allow to invoke polkit agent even if STDIN is not a tty

### LINKED ISSUE #35018: run0: headless piping into run0 authentication error, "Failed to start transient service unit: Interactive authentication required."
### systemd version the issue has been seen with

256.7

### Used distribution

Arch Linux

### Linux kernel version used

6.11.6-arch1-1

### CPU architectures issue was seen on

x86_64

### Component

systemd-run

### Expected behaviour you didn't see

I expect that if I ssh into my headless server and do `echo "ls -al /root" | run0 bash` as a non-root user, then it should work just like it would if I replaced `run0` with `sudo`, i.e. I should get a prompt to enter my password and then get a listing of the /root directory

### Unexpected behaviour you saw

On my headless server as user in wheel group (unexpected behavior):
```
$ echo "ls -al /root" | run0 bash
Failed to start transient service unit: Interactive authentication required.
```

### Expected behavior you saw

On my headless server as root user (expected behavior):
```
# echo "ls -al /root" | run0 bash
total 0
drwxr-x--- 1 root root   0 Nov  4 18:20 .
drwxr-xr-x 1 root root 122 Nov  4 18:20 ..
```

In the console on my GNOME desktop computer as a user in wheel group (expected behavior):
```
$ echo "ls -al /root" | run0 bash
# UI prompt window appears now asking for my password, and after entering it
total 0
drwxr-x--- 1 root root   258 Oct 30 00:40 .
drwxr-xr-x 1 root root   136 May  8 20:52 ..
```

On my headless server as a user in wheel group using `sudo` in place of `run0` (expected behavior):
```
$ echo "ls -al /root" | sudo bash
[sudo] password for user:
total 0
drwxr-x--- 1 root root   0 Nov  4 18:20 .
drwxr-xr-x 1 root root 122 Nov  4 18:20 ..
```

On my headless server as a user in wheel group, running the command directly instead of piping it into bash (expected behavior):
```
$ run0 ls -al /root
==== AUTHENTICATING FOR org.freedesktop.systemd1.manage-units ====
Authentication is required to manage system services or other units.
Authenticating as: user
Password: 
==== AUTHENTICATION COMPLETE ====
total 0
drwxr-x--- 1 root root   0 Nov  4 18:20 .
drwxr-xr-x 1 root root 122 Nov  4 18:20 ..
```

### Steps to reproduce the problem

1. Use Arch Linux (doubtful that distro matters, though I haven't tested others) without a desktop environment installed
2. run `echo "ls -al /root" | run0 bash` as a non-root user

### Additional program output to the terminal or log subsystem illustrating the issue

_No response_

### PR DESCRIPTION
**Pull request by yuwata**
Closes #35018.

### CODE REVIEW

File: src/shared/exec-util.c [hunk 1]
```diff
@@ -567,17 +566,18 @@ int _fork_agent(const char *name, const int except[], size_t n_except, pid_t *re
...
-        if (!stdout_is_tty || !stderr_is_tty) {
+        if (!stdin_is_tty || !stdout_is_tty || !stderr_is_tty) {
                 int fd;
 
                 /* Detach from stdout/stderr and reopen /dev/tty for them. This is important to ensure that
                  * when systemctl is started via popen() or a similar call that expects to read EOF we
                  * actually do generate EOF and not delay this indefinitely by keeping an unused copy of
                  * stdin around. */
-                fd = open("/dev/tty", O_WRONLY);
+                fd = open("/dev/tty", stdin_is_tty ? O_WRONLY : O_RDWR);
```
Comment (poettering) [inline src/shared/exec-util.c]: shouldnt we also use O_RDONLY here if only input is needed? just to be fully correct?

File: src/shared/exec-util.c [hunk 2]
```diff
@@ -588,13 +588,18 @@ int _fork_agent(const char *name, const int except[], size_t n_except, pid_t *re
                          * connected to a TTY. That's a weird setup, but let's handle it gracefully: let's
```
Comment (poettering) [inline src/shared/exec-util.c]: this comment needs some minor updating now

File: src/shared/polkit-agent.c [hunk 3]
```diff
@@ -31,10 +32,6 @@ int polkit_agent_open(void) {
         if (geteuid() == 0)
                 return 0;
 
-        /* We check STDIN here, not STDOUT, since this is about input, not output */
-        if (!isatty_safe(STDIN_FILENO))
-                return 0;
-
         /* Also check if we have a controlling terminal. If not (ENXIO here), we aren't actually invoked
```
Comment (poettering) [inline src/shared/polkit-agent.c]: The "also" in this comment doesn#t make too much sense now anymore

File: src/shared/exec-util.c [hunk 4]
```diff
@@ -567,17 +563,18 @@ int _fork_agent(const char *name, const int except[], size_t n_except, pid_t *re
 
         /* In the child: */
 
-        stdout_is_tty = isatty_safe(STDOUT_FILENO);
-        stderr_is_tty = isatty_safe(STDERR_FILENO);
+        bool stdin_is_tty = isatty_safe(STDIN_FILENO);
+        bool stdout_is_tty = isatty_safe(STDOUT_FILENO);
+        bool stderr_is_tty = isatty_safe(STDERR_FILENO);
```
Comment (YHNdnzj) [inline src/shared/exec-util.c]: Just use one `bool` with multiple comma-delimited vars?

**Review by YHNdnzj — APPROVED** [review]

**Comment by yuwata** [comment]
Updated. Change from the last review is slightly big. Hence, dropping the green label. PTAL!

**Comment by bluca** [comment]
If you want this to be backported, please rebase and reorder so that the bugfix is first. If it's only for 258 and not for backports then it's ok as-is.

**Comment by yuwata** [comment]
> If you want this to be backported, please rebase and reorder so that the bugfix is first. If it's only for 258 and not for backports then it's ok as-is.

OK. Done.

File: src/shared/exec-util.c [hunk 5]
```diff
@@ -543,6 +543,23 @@ int fexecve_or_execve(int executable_fd, const char *executable, char *const arg
         return -errno;
 }
 
+int can_fork_agent(void) {
```
Comment (poettering) [inline src/shared/exec-util.c]: I think "shall" rather than "can" would be better.

**Pull request merged**